providers/implementations/keymgmt/hybrid_kmgmt.c

```diff
@@ -0,0 +1,667 @@
+                                          OSSL_PKEY_PARAM_SECURITY_BITS);
+            if (t == 0 || !OSSL_PARAM_set_uint32(p, t))
+                return 0;
+            break;
+
+        case PIDX_PKEY_PARAM_BITS:
+            t = accumulate_ctx_uint32_params(ctx->num_algs, ctx->ctxs,
+                                             OSSL_PKEY_PARAM_BITS, NULL);
+            if (t == 0 || !OSSL_PARAM_set_uint32(p, t))
+                return 0;
+            break;
+
+        case PIDX_PKEY_PARAM_MAX_SIZE:
+            t = ossl_hybrid_get_key_length_from_ctx(ctx);
+            if (t == 0 || !OSSL_PARAM_set_uint32(p, t))
+                return 0;
+            break;
+
+        case PIDX_PKEY_PARAM_ENCODED_PUBLIC_KEY:
+            t = ossl_hybrid_get_key_length_from_ctx(ctx);
+            if (t > p->data_size) {
+                p->data_size = t;
+                ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH,
+                               "length must be %u bytes, have %u bytes",
+                               t, p->data_size);
+                return 0;
+            }
+            p->data_size = t;
+            q = p->data;
+            *prms = *p;
+            for (i = 0; i < ctx->num_algs; i++) {
+                prms->data = q;
+                prms->data_size = ctx->key_lengths[i];
+                q += ctx->key_lengths[i];
+                if (!EVP_PKEY_CTX_get_params(ctx->ctxs[i], prms))
+                    return 0;
+            }
+            break;
+
+        default:
+            if (!ctx_subalg_get_param(ctx->num_algs, ctx->ctxs, p))
+                return 0;
+            break;
+        }
+    }
+    return 1;
+}
+
+const OSSL_PARAM *ossl_hybrid_gettable_ctx_params(HYBRID_PKEY_CTX *vctx,
+                                                  ossl_unused void *provctx)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    const OSSL_PARAM *pary[MAX_HYBRID_ALGS + 1];
+    unsigned int i;
+    static const OSSL_PARAM hybrid_gen_gettable_params[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, NULL, 0),
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_BITS, NULL),
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_SECURITY_BITS, NULL),
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_MAX_SIZE, NULL),
+        OSSL_PARAM_END
+    };
+
+    if (ctx->gettable == NULL) {
+        for (i = 0; i < ctx->num_algs; i++)
+            pary[i] = EVP_PKEY_CTX_gettable_params(ctx->ctxs[i]);
+        pary[i] = hybrid_gen_gettable_params;
+        ctx->gettable = ossl_param_merge(ctx->num_algs + 1, pary);
+    }
+    return ctx->gettable;
+}
+
+int ossl_hybrid_set_ctx_params(HYBRID_PKEY_CTX *ctx, const OSSL_PARAM params[])
+{
+    char *str = NULL;
+    const OSSL_PARAM *p;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    if (!ctx_subalg_set_param(ctx->num_algs, ctx->ctxs, params))
+        return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
```

Comment:
Maybe propq should be done first?

Reply:
It is done first in the sub algs.  This is keeping a local copy.